nncf/onnx/graph/nncf_graph_builder.py

```diff
@@ -21,11 +21,183 @@
 
 
+class ONNXConstantLayerAttributes(BaseLayerAttributes):
+    """
+    This class stores extended attributes of modules/layers for the algorithms.
+    """
+
+    def __init__(self, weight_attrs: Dict[int, Dict], bias_attrs: Dict[int, Dict]):
+        """
+        :param const_attrs: Map of weights port ID to corresponding const attributes.
+        """
+        self.weight_attrs = weight_attrs
+        self.bias_attrs = bias_attrs
+
+    def get_weight_port_ids(self) -> List[int]:
+        """
+        Returns indices of input ports corresponding to the constant nodes.
+
+        :returns: List of input port indices with constants.
+        """
+        return list(self.weight_attrs.keys())
+
+    def get_bias_port_ids(self) -> List[int]:
+        """
+        Returns indices of input ports corresponding to the constant nodes.
+
+        :returns: List of input port indices with constants.
+        """
+        return list(self.bias_attrs.keys())
+
+
+def _get_weight_edge_name(onnx_graph: ONNXGraph, node: onnx.NodeProto, port_id: int) -> Optional[str]:
+    """
+    Finds an edge associated with a node weight on a input port_id.
+    If an edge was not found then returns None.
+    If edge associated with a weight was found then returns edge name.
+
+    :param onnx_graph: ONNXGraph.
+    :param node: Node.
+    :param int port_id: Port id on which a weight edge is seeking.
+    :return: Edge name associated with a weight.
+    """
+
+    # There are several cases
+    # (Constant) -> (Operation)
+    # (Identity) -> (Operation)
+    # (Reshape) -> (Operation)
+    # (Transpose) -> (Operation)
+    # (Constant) -> (QuantizeLinear) -> (DequantizeLinear) -> (Operation)
+    # (Constant) -> (QuantizeLinear) -> (DequantizeLinear) -> (Operation)
+
+    #  We need properly find the constant node. So we start with
+    # Operation and traverse up until the constant node is not found.
+
+    # If Constant on the port
+    WEIGHT_CONSUMING_NODES = (
+        ONNXConstantMetatype.get_all_aliases()
+        + ONNXIdentityMetatype.get_all_aliases()
+        + ONNXReshapeMetatype.get_all_aliases()
+        + ONNXTransposeMetatype.get_all_aliases()
+        + ONNXQuantizeLinearMetatype.get_all_aliases()
+    )
+    ALLOWED_OP_TYPES = WEIGHT_CONSUMING_NODES + ONNXDequantizeLinearMetatype.get_all_aliases()
+
+    weight_index = 0
+    if onnx_graph.has_initializer(node.input[port_id]):
+        return node.input[port_id]
+    queue = onnx_graph.get_parent(node, port_id)
+    while queue:
+        node = queue.pop()
+        if node.op_type not in ALLOWED_OP_TYPES:
+            return None
+        if node.op_type in WEIGHT_CONSUMING_NODES:
+            if onnx_graph.has_initializer(node.input[weight_index]):
+                return node.input[weight_index]
+            return None
+        queue.extend(onnx_graph.get_parent(node, 0))
+    return None
+
+
+def _get_layer_name(weight_edge_names: List[str]) -> str:
```

Comment:
Want to discuss the layer name attribute of NNCFNode in case the node has several shared weights.

Reply:
Done